rust/src/traits/samplers/psrn/test.rs

```diff
@@ -1,32 +1,106 @@
 use super::*;
 
-#[test]
-fn test_sample_gumbel_interval_progression() -> Fallible<()> {
-    let mut gumbel = GumbelPSRN::new(RBig::ZERO, RBig::ONE);
-    for _ in 0..10 {
-        println!(
-            "{:?}, {:?}, {}",
-            gumbel.value::<Down>()?.to_f64(),
-            gumbel.value::<Up>()?.to_f64(),
-            gumbel.precision
-        );
-        gumbel.refine()?;
+/// Conduct a Kolmogorov-Smirnov (KS) test.
+///
+/// Since the critical values are difficult to compute,
+/// this function hardcodes the critical value corresponding to a p-value of 1e-6 when 1000 samples are taken.
+///
+/// Assuming the samples are draws from the distribution specified by the cdf,
+/// then the p-value is the false discovery rate,
+/// or chance of this test failing even when the data is a sample from the distribution.
+pub fn kolmogorov_smirnov(mut samples: [f64; 1000], cdf: impl Fn(f64) -> f64) -> Fallible<()> {
+    // first, compute the test statistic. For a one-sample KS test,
+    // this is the greatest distance between the empirical CDF of the samples and the expected CDF.
+    samples.sort_by(|a, b| a.total_cmp(b));
+
+    let n = samples.len() as f64;
+    let statistic = samples
+        .into_iter()
+        .enumerate()
+        .map(|(i, s)| {
+            let empirical_cdf = i as f64 / n;
+            let idealized_cdf = cdf(s);
+            (empirical_cdf - idealized_cdf).abs()
+        })
+        .max_by(|a, b| a.total_cmp(b))
+        .unwrap();
+
+    // The KS-test is nonparametric,
+    // so the critical value only changes in response to the number of samples (hardcoded at 1000),
+    // not the distribution.
+    //
+    // The p-value corresponds to the mass of the tail of the KS distribution beyond the critical value.
+    // The mass of the tail is the complement of the cumulative distribution function,
+    // which is also called the survival function.
+    // The inverse of the survival function `isf` tells us the critical value corresponding to a given mass of the tail.
+    //
+    // We therefore derive the critical value via the inverse survival function of the two-sided, one-sample KS distribution:
+    // ```python
+    // from scipy.stats import kstwo
+    // CRIT_VALUE = kstwo(n=1000).isf(1e-6)
+    // ```
+    static CRIT_VALUE: f64 = 0.08494641956324511;
+    if statistic > CRIT_VALUE {
+        return fallible!(FailedFunction, "Statistic ({statistic}) exceeds critical value ({CRIT_VALUE})! This indicates that the data is not sampled from the same distribution specified by the cdf.");
```

Comment:
```suggestion
        return fallible!(FailedFunction, "Statistic ({statistic}) exceeds critical value ({CRIT_VALUE})! This indicates that the data is not sampled from the same distribution specified by the cdf (with probability 1 - 1e-6).");
```
The message is fine as-is, but just to check my understanding: There is a very small chance that this will give a false error?

Reply:
Indeed. Updated the message.